docs: Add porting guide for PSA targets

urutva opened the pull request:
The porting guide describes the process of adding new PSA targets to
Mbed OS. The focus is more on target configurations, build and
validation.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Edit by Irit: preview (only on VPN): https://os.mbed.com/docs/mbed-os/adding-psa-targets/porting/porting-security.html

urutva committed:
docs: Add porting guide for PSA targets

The porting guide describes the process of adding new PSA targets to
Mbed OS. The focus is more on target configurations, build and
validation.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

devran01 commented:
@iriark01 @Patater Please review.

iriark01 commented:
Should I wait for Jaeden's review, then?

iriark01 committed:
Pre-review: getting it to build

devran01 commented:
@Patater Are you ok with Irit starting the review?

Patater commented:
I've reviewed this already in the Mbed OS PR. Don't wait for me.

iriark01 committed:
Not finished yet - just having computer issues

iriark01 committed:
Ready for engineering review

Co-authored-by: Irit Arkin <irit.arkin@arm.com>

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+# PSA targets in Mbed OS
```
```suggestion
# Porting PSA targets
```

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+
+### Inheritance
+
+The Mbed OS build tools have limited support for multiple inheritance. Because of this, a PSA target must inherit from the generic PSA target corresponding to the new target's architecture. The only exception is when the PSA target has to inherit from one of its family targets instead of the generic target `Target`<!--should that be `psa_target`? If so, that should be in the first sentence, not h ...
```
```suggestion
The Mbed OS build tools have limited support for multiple inheritance. Because of this, a PSA target must inherit from the generic PSA target corresponding to the new target's architecture. The only exception is when the PSA target has to inherit from one of its family targets (one of `PSA_V7_M`, `PSA_DUAL_CORE` and `PSA_V8_M`) instead of the generic target. We cover exception cases in each target section in this document.
```

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+
+    The supported values are `ARMCLANG` and `GNUARM`.
+* `tfm_delivery_dir`: The directory to which TF-M binaries will be copied.
+* `TFM_OUTPUT_EXT`: Optional attribute that indicates the output extension of the TF-M secure binary.<!--how come only this one is all caps?-->
```
```suggestion
* `tfm_output_ext`: Optional attribute that indicates the output extension of the TF-M secure binary.
```

urutva replied:
Unfortunately, we can't change it here. The code contains capitalised text.

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+
+To add or remove a feature, use `feature_`[add/remove].
+
+By default, a TF-M build generates a `bin` file. If the target requires a `hex` file, you need to add the attribute `"TFM_OUTPUT_EXT": "hex"` to the target definition. The build script will convert `bin` to `hex`, then copy the `hex` to `tfm_delivery_dir`.
```
```suggestion
By default, a TF-M build generates a `bin` file. If the target requires a `hex` file, you need to add the attribute `"tfm_output_ext": "hex"` to the target definition. The build script will convert `bin` to `hex`, then copy the `hex` to `tfm_delivery_dir`.
```

iriark01 replied:
Ah, I misunderstood you - I thought you meant it was an error in the docs, not the code ;)

urutva replied:
Unfortunately, we can't change it here. The code contains capitalised text.

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+     The supported values are `ARMCLANG` and `GNUARM`.
+
+* `tfm_delivery_dir`: The directory to which TF-M binary will be copied.
+* `TFM_OUTPUT_EXT`: Optional attribute that indicates the output extension of the TF-M secure binary.
```
```suggestion
* `tfm_output_ext`: Optional attribute that indicates the output extension of the TF-M secure binary.
```

urutva replied:
Unfortunately, we can't change it here. The code contains capitalised text.

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+
+To add or remove a feature, use `feature_`[add/remove].
+
+By default, a TF-M build generates a `bin` file. If the target requires a `hex` file, you need to add the attribute `"TFM_OUTPUT_EXT": "hex"` to the target definition. The build script will convert `bin` to `hex`, then copy the `hex` to `tfm_delivery_dir`. You must also update `secure_image_filename` to match the new file extension.
```
```suggestion
By default, a TF-M build generates a `bin` file. If the target requires a `hex` file, you need to add the attribute `"tfm_output_ext": "hex"` to the target definition. The build script will convert `bin` to `hex`, then copy the `hex` to `tfm_delivery_dir`. You must also update `secure_image_filename` to match the new file extension.
```

urutva replied:
Unfortunately, we can't change it here. The code contains capitalised text.

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+
+## Build and validation
+
+For dual-core and Armv8-M PSA targets, TF-M runs on the SPE and provides PSA services<!--I feel like we've said this at least twice already-->. The Python script `build_tfm.py` automates building TF-M and copying the TF-M binary to the location defined by the target attribute `tfm_delivery_dir`. Another Python script, `build_psa_compliance.py`, automates
```
```suggestion
The Python script `build_tfm.py` automates building TF-M and copying the TF-M binary to the location defined by the target attribute `tfm_delivery_dir`. Another Python script, `build_psa_compliance.py`, automates
```

devran01 replied:
sure, it’s just a pre-text to define (again) SPE and NSPE and then say how to build SPE. Can be removed.

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+To build TF-M and regression tests:
+
+1. Clone the [mbed-os-tf-m-regression-tests](https://github.com/ARMmbed/mbed-os-tf-m-regression-tests) repository.
+1. Update `mbed-os.lib` to the version<!--version or *location*? I assume you mean their local copy, basically--> of Mbed OS that contains new target support, including the `targets.json` changes described in this document.
```
```suggestion
1. Update `mbed-os.lib` to the copy of Mbed OS that contains new target support, including the `targets.json` changes described in this document.
```

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+
+    <span class="notes">**Note:** Use this `ConfigRegressionIPC.cmake` only to run regression tests. The secure binary copied to Mbed OS (when you are ready to merge your new target) must be generated using the config option `ConfigCoreIPC.cmake`.</span>
+
+1. Ensure the TF-M binary was copied to the location defined by the target attribute `tfm_delivery_dir`.<!--this is assuming we really generated a binary in the previous step - I was making an assumption-->
```
```suggestion
1. Ensure the TF-M binary was copied to the location defined by the target attribute `tfm_delivery_dir`. If no binary was created or copied, please raise an [issue on the Mbed OS repository](https://github.com/ARMmbed/mbed-os).
```

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+   If everything looks good, push the changes to your fork of Mbed OS.
+1. Create an Mbed OS pull request against [https://github.com/ARMmbed/mbed-os](https://github.com/ARMmbed/mbed-os).
+
+   <!--should we include a link to anything in the contributing guide, or does that not matter within the context of the porting guide?-->
```
```suggestion
<span class="tips">**Tip:** See the [contributing guide](../contributing/index.html) for more information about pull requests.</span>
```

urutva commented on docs/porting/security/psa_targets_in_mbed-os.md:
````diff
@@ -0,0 +1,561 @@
+```
+
+Note the config options `macros_add`, `extra_labels_add` and `device_has_add`. To add or remove `macros`, `extra_labels` or target capabilities, a PSA target definition **must** use [macros/extra_labels/device_has]`_add` or [macros/extra_labels/device_has]`_remove` (not `macros`, `extra_labels` or `device_has`).
+
````
You moved the links from macros , extra_labels , device_has  and Features  to a new section. Wouldn’t it be easier for the user, if those would have been hyperlinks rather than links in a different section?

iriark01 replied:
But we had it three or four times in the docs. That can make people try to figure out whether there's a difference between those times, which is a waste of effort, or - worse - can make them skim the end of each section on the assumption that more of it repeats.

devran01 replied:
Irit: But we had it three or four times in the docs. That can make people try ti figure out whether there's a different between those times, which is a waste of effort, or - worse - can make them skim the end of each section on the assumption that more of it repeats

urutva replied:
@iriark01 Agreed, that makes sense.

urutva commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+* `PSA_DUAL_CORE`: Dual-core PSA generic target. A dual-core PSA target has at least two cores that are either Armv7-M or Armv6-M. One core will be used for the SPE and another for the NSPE. Hardware isolation between the cores enables PSA compliance. On dual-core targets, TF-M
+runs on the SPE and provides PSA services, and Mbed OS runs on the NSPE.
+* `PSA_V8_M`: Armv8-M generic PSA target, which employs hardware to isolate the NSPE from the SPE. TF-M runs on the SPE and provides PSA services, and Mbed OS runs on the NSPE.
+<!--there are two ways of describing things in this list: chip-generic-psa and chip-psa-generic. Is that on purpose?-->
```
Can you elaborate bit more on this? The list defines new PSA targets categorised based on the version of Arm architecture it supports.

iriark01 replied:
I meant that you describe them in two different orders, and I wondered if they should be consistent

devran01 replied:
Irit: I meant that you describe them in two different orders, and I wondered if they should be consistent

urutva replied:
@iriark01 Hardware isolation is the differentiating factor, all three targets cover that, but `PSA_DUAL_CORE` needs explanation on the hardware setup. Hence the preamble. I can't think of other way to describe it. Any input is welcome :)

iriark01 replied:
I mean specifically the text of `PSA generic target` v `generic PSA target`

urutva replied:
Ah totally missed that. It should be `generic PSA target`. I'll fix it. :)

urutva replied:
Fixed.

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+
+**Tip:** When you run the Python script `build_tfm.py` without any options, TF-M is built for all supported targets and the secure binary (TF-M) for each target is copied to the location defined by the target attribute `tfm_delivery_dir`. When you use the `--commit` option, a new Mbed OS commit is created with the new or modified TF-M binaries and `features/FEATURE_PSA/TARGET_TFM/VERSION.txt`.
+
+Example:<!--isn't both the example and the text above it really just part of step 2 above?-->
```
I think the explanation should be in step 2 or before it, not after. And the example should be removed or be before the one with -m and -t. Otherwise this can confuse users - it's two takes on a topic

iriark01 replied:
@devran01 this for some reason isn't showing up in the text review, but still needs a decision

urutva replied:
@iriark01 It's already fixed. Sorry I forgot to comment :)

urutva replied:
@iriark01 missed the example. I'm happy to remove it.

urutva replied:
Removed the example.

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+
+By default, a TF-M build generates a `bin` file. If the target requires a `hex` file, you need to add the attribute `"TFM_OUTPUT_EXT": "hex"` to the target definition. The build script will convert `bin` to `hex`, then copy the `hex` to `tfm_delivery_dir`.
+
+This dual-core PSA target doesn't inherit from `PSA_DUAL_CORE` because it has to inherit from one of its family targets<!--I'm still not sure why I have to inherit from a family target-->. Hence, we have added additional attributes:
```
I still want to understand under what circumstances a target has to inherit from its family. Why exactly does it happen.

urutva replied:
@iriark01 Mbed OS config system supports inheritance. The idea behind the concept is that, target configs can be defined for a parent target (MCU family: example - https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json#L6712 and https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json#L6776). This allows child targets config to be minimal and avoids duplication (target board: example - https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json#L6913). When this happens, the child target cannot inherit from PSA generic target because of limitations of Mbed OS build tools.

iriark01 replied:
So is the constraint that it can't inherit from both at once?

urutva replied:
Yeah, multiple inheritance support is very limited.

iriark01 replied:
Okay, I added an attempt at clarification for line 32

devran01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,561 @@
+
+**Tip:** When you run the Python script `build_tfm.py` without any options, TF-M is built for all supported targets and the secure binary (TF-M) for each target is copied to the location defined by the target attribute `tfm_delivery_dir`. When you use the `--commit` option, a new Mbed OS commit is created with the new or modified TF-M binaries and `features/FEATURE_PSA/TARGET_TFM/VERSION.txt`.
+
+Example:<!--isn't both the example and the text above it really just part of step 2 above?-->
```
I think the explanation should be in step 2 or before it, not after. And the example should be removed or be before the one with -m and -t. Otherwise this can confuse users - it's two takes on a topic

urutva committed:
docs: cleanup

Remove reference to `MBEDTLS_PSA_HAS_ITS_IO` and
`MBEDTLS_USE_PSA_CRYPTO` which are not needed anymore.

Remove section which describes supported options for python scrips.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
docs: Fix review comments

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,495 @@
+
+### Inheritance
+
+The Mbed OS build tools have limited support for multiple inheritance. Because of this, a PSA target must inherit from the generic PSA target corresponding to the new target's architecture. The only exception is when the PSA target has to inherit from one of its family targets (one of `PSA_V7_M`, `PSA_DUAL_CORE` and `PSA_V8_M`) instead of the generic target. We cover exception cases in each targe ...
```
```suggestion
The Mbed OS build tools have limited support for multiple inheritance. Because of this, a PSA target must inherit from the generic PSA target corresponding to the new target's architecture *or* from one of its family targets (one of `PSA_V7_M`, `PSA_DUAL_CORE` and `PSA_V8_M`), but cannot inherit from both the generic and family targets at once. We cover family inheritance cases in each target section in this document.
```

iriark01 replied:
@devran01 how's that?

urutva replied:
@iriark01 Much better. One change though,
``` Because of this, a PSA target must inherit from the generic PSA target corresponding to the new target's architecture (one of `PSA_V7_M`, `PSA_DUAL_CORE` and `PSA_V8_M`) *or* from one of its family targets, but cannot inherit from both the generic and family targets at once.```
If you are happy with this then I'll update it.

iriark01 replied:
Ah, I didn't realise I had those target names switched! Sure, go ahead. Thanks.

urutva replied:
Fixed.

devran01 commented:
@iriark01 Please do not merge this PR till https://github.com/ARMmbed/mbed-os/pull/12955 is merged.

Patater commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,489 @@
+# Porting PSA targets
+
+This document describes the process of adding new PSA targets to Mbed OS, focusing on target configurations, build and validation. Mbed OS relies on [TF-M](https://git.trustedfirmware.org/trusted-firmware-m.git/tree/) for PSA Services and SPM, so the PSA target you are adding to Mbed OS **must** already be supported by TF-M.
```
We wouldn't expect readers to know what SPM is. Instead of "PSA Services and SPM", could we instead say "for its implementation of PSA"?

iriark01 replied:
This is because the glossary is after the first paragraph; Dev originally put it before the content, but I found it distracting.

Patater replied:
I don't think we need to explain SPM either. That's an unnecessary detail for porting targets and we don't need to mention SPM here or in any glossary.

urutva replied:
fixed

Patater commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,489 @@
+
+    The supported values are `ARMCLANG` and `GNUARM`.
+* `tfm_delivery_dir`: The directory to which TF-M binaries will be copied.
+* `TFM_OUTPUT_EXT`: Optional attribute that indicates the output extension of the TF-M secure binary.<!--how come only this one is all caps?-->
```
Do we want to remove the `<!--how come only this one is all caps?-->` or other `<!--` text from the markdown for resolved issues?

iriark01 replied:
Yes, though if we miss one it's not the end of the world as it doesn't get rendered.

Patater commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,489 @@
+
+## Enabling PSA at application level
+
+Having an entropy source is crucial for Mbed TLS and Mbed Crypto. The [entropy source porting document](../porting/entropy-sources.html) discusses why and how to add an entropy source.
```
Replace "Mbed Crypto" with "PSA". "Mbed Crypto" doesn't exist anymore (was rolled back into Mbed TLS).

urutva replied:
fixed

Patater commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,489 @@
+
+Having an entropy source is crucial for Mbed TLS and Mbed Crypto. The [entropy source porting document](../porting/entropy-sources.html) discusses why and how to add an entropy source.
+
+If your target doesn't have a True Random Number Generator (TRNG), configure it as a non-PSA target in `targets/targets.json`. In this scenario, if an application wants to use the target as a PAS target, then it is the application's responsibility to provide an entropy source and mark that target as a PSA target at the application level. To enable PSA for a target from an application, use the con ...
```
Change "PAS target" to "PSA target"

urutva replied:
fixed

Patater commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,489 @@
+1. Edit `main.cpp` and set:
+
+    - `RUN_PSA_COMPLIANCE_TESTS` to `1`.
+    - `RUN_REGRESSION_TESTS` to `0`.
```
This isn't how to select the tests anymore. We can use mbed_app.json.

urutva replied:
fixed

Patater commented on docs/porting/security/psa_targets_in_mbed-os.md:
````diff
@@ -0,0 +1,489 @@
+    python3 build_tfm.py -m <new target> -t <toolchain> --commit
+    ```
+
+    The command builds TF-M with the config `ConfigCoreIPC.cmake` ,copies the TF-M binary to the location defined by the target attribute `tfm_delivery_dir`, and commits the changes to Mbed OS.
````
Should we make this change? or is the space where it is because of the particular markdown tool we are using?
```diff
- the config `ConfigCoreIPC.cmake` ,copies the
+ the config `ConfigCoreIPC.cmake`, copies the
```

iriark01 replied:
It's just a typo

urutva replied:
fixed

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,489 @@
+
+By default, a TF-M build generates a `bin` file. If the target requires a `hex` file, you need to add the attribute `"TFM_OUTPUT_EXT": "hex"` to the target definition. The build script will convert `bin` to `hex`, then copy the `hex` to `tfm_delivery_dir`.
+
+This dual-core PSA target doesn't inherit from `PSA_DUAL_CORE` because it has to inherit from one of its family targets<!--I'm still not sure why I have to inherit from a family target-->. Hence, we have added additional attributes:
```
```suggestion
This dual-core PSA target doesn't inherit from `PSA_DUAL_CORE` because it has to inherit from one of its family targets. Hence, we have added additional attributes:
```

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,489 @@
+
+    The supported values are `ARMCLANG` and `GNUARM`.
+* `tfm_delivery_dir`: The directory to which TF-M binaries will be copied.
+* `TFM_OUTPUT_EXT`: Optional attribute that indicates the output extension of the TF-M secure binary.<!--how come only this one is all caps?-->
```
```suggestion
* `TFM_OUTPUT_EXT`: Optional attribute that indicates the output extension of the TF-M secure binary.
```

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,489 @@
+* `PSA_DUAL_CORE`: Dual-core generic PSA target. A dual-core PSA target has at least two cores that are either Armv7-M or Armv6-M. One core will be used for the SPE and another for the NSPE. Hardware isolation between the cores enables PSA compliance. On dual-core targets, TF-M
+runs on the SPE and provides PSA services, and Mbed OS runs on the NSPE.
+* `PSA_V8_M`: Armv8-M generic PSA target, which employs hardware to isolate the NSPE from the SPE. TF-M runs on the SPE and provides PSA services, and Mbed OS runs on the NSPE.
+<!--there are two ways of describing things in this list: chip-generic-psa and chip-psa-generic. Is that on purpose?-->
```
```suggestion
```

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,488 @@
+
+## Enabling PSA at application level
+
+Having an entropy source is crucial for Mbed TLS and Mbed Crypto. The [entropy source porting document](../porting/entropy-sources.html) discusses why and how to add an entropy source.
```
```suggestion
Having an entropy source is crucial for Mbed TLS and PSA. The [entropy source porting document](../porting/entropy-sources.html) discusses why and how to add an entropy source.
```

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
```diff
@@ -0,0 +1,488 @@
+
+Having an entropy source is crucial for Mbed TLS and Mbed Crypto. The [entropy source porting document](../porting/entropy-sources.html) discusses why and how to add an entropy source.
+
+If your target doesn't have a True Random Number Generator (TRNG), configure it as a non-PSA target in `targets/targets.json`. In this scenario, if an application wants to use the target as a PAS target, then it is the application's responsibility to provide an entropy source and mark that target as a PSA target at the application level. To enable PSA for a target from an application, use the con ...
```
```suggestion
If your target doesn't have a True Random Number Generator (TRNG), configure it as a non-PSA target in `targets/targets.json`. In this scenario, if an application wants to use the target as a PSA target, then it is the application's responsibility to provide an entropy source and mark that target as a PSA target at the application level. To enable PSA for a target from an application, use the config option [target_overrides](,,/reference/configuration.html).
```

iriark01 commented on docs/porting/security/psa_targets_in_mbed-os.md:
````diff
@@ -0,0 +1,488 @@
+    python3 build_tfm.py -m <new target> -t <toolchain> --commit
+    ```
+
+    The command builds TF-M with the config `ConfigCoreIPC.cmake` ,copies the TF-M binary to the location defined by the target attribute `tfm_delivery_dir`, and commits the changes to Mbed OS.
````
```suggestion
    The command builds TF-M with the config `ConfigCoreIPC.cmake`, copies the TF-M binary to the location defined by the target attribute `tfm_delivery_dir`, and commits the changes to Mbed OS.
```

Patater reviewed (approved):
LGTM

iriark01 committed:
Update psa_targets_in_mbed-os.md
